modules/AlphaWalletFoundation/AlphaWalletFoundation/EtherClient/TransactionSigning.swift

```diff
@@ -3,70 +3,210 @@
+        }
+    }
+
+    public func rlpEncodedHash(transaction: UnsignedTransaction) throws -> Data {
+        switch transaction.gasPrice {
+        case .legacy(let gasPrice):
+            return try rlpEncodeForLegacyGasPriceHash(transaction: transaction, legacyGasPrice: gasPrice)
+        case .eip1559(let maxFeePerGas, let maxPriorityFeePerGas):
+            return try rlpEncodeForEip1559Hash(transaction: transaction, maxFeePerGas: maxFeePerGas, maxPriorityFeePerGas: maxPriorityFeePerGas)
         }
+    }
 
+    private func rlpEncodeForLegacyGasPriceHash(transaction: UnsignedTransaction, legacyGasPrice: BigUInt) throws -> Data {
         let values: [Any] = [
             transaction.nonce,
-            transaction.gasPrice,
+            legacyGasPrice,
             transaction.gasLimit,
             transaction.to?.data ?? Data(),
             transaction.value,
             transaction.data,
             transaction.server.chainID, 0, 0,
         ]
 
-        guard let data = rlpHash(values) else {
-            throw AnyError.invalid
-        }
+        guard let data = rlpHash(values) else { throw SignerError.rplEncodeFailure }
         return data
     }
 
-    public func values(signature: Data) -> (r: BigInt, s: BigInt, v: BigInt) {
-        let (r, s, v) = HomesteadSigner().values(signature: signature)
-        let newV: BigInt
-        if server.chainID != 0 {
-            newV = BigInt(signature[64]) + 35 + BigInt(server.chainID) + BigInt(server.chainID)
-        } else {
-            newV = v
+    private func rlpEncodeForEip1559Hash(transaction: UnsignedTransaction, maxFeePerGas: BigUInt, maxPriorityFeePerGas: BigUInt) throws -> Data {
+        let values: [Any] = [
+            transaction.server.chainID,
+            transaction.nonce,
+            maxPriorityFeePerGas,
+            maxFeePerGas,
+            transaction.gasLimit,
+            transaction.to?.data ?? Data(),
+            transaction.value,
+            transaction.data,
+            [],
+        ]
+
+        guard let data = RLP.encode(values) else { throw SignerError.rplEncodeFailure }
+        return Data(bytes: ([0x02] + data.bytes).sha3(.keccak256))
+    }
+
+    public func signature(transaction: UnsignedTransaction, signatureData: Data) -> EthereumSignature {
+        switch transaction.gasPrice {
+        case .legacy:
+            return signatureLegacy(from: signatureData)
+        case .eip1559:
+            return signatureEip1559(from: signatureData)
         }
-        return (r, s, newV)
+    }
+
+    private func signatureLegacy(from data: Data) -> EthereumSignature {
+        return EthereumSignature(
+            v: server.chainID == 0 ? data[64] + EthereumSigner.vitaliklizeConstant : data[64] + 35 + UInt8(server.chainID * 2),
+            r: data[..<32].bytes,
+            s: data[32..<64].bytes)
+    }
+
+    private func signatureEip1559(from data: Data) -> EthereumSignature {
+        return EthereumSignature(
+            v: data[64],
```

Comment:
Should this require ` + EthereumSigner.vitaliklizeConstant` and chainID for `v` too, or does EIP-1559 just take the parity bit as-is??